demos/text_detection_demo/src/text_recognition.cpp

```diff
@@ -53,3 +76,83 @@ std::string CTCGreedyDecoder(const std::vector<float> &data, const std::string&
     }
     return res;
 }
+
+std::string CTCBeamSearchDecoder(const std::vector<float> &data, const std::string& alphabet, char pad_symbol, double *conf, int bandwidth) {
+    std::string res = "";
+
+    const int num_classes = alphabet.length();
+
+    std::vector<BeamElement> curr;
+    std::vector<BeamElement> last;
+
+    last.push_back(BeamElement{"", 1.f, 0.f});
+
+    for (std::vector<float>::const_iterator it = data.begin(); it != data.end(); it += num_classes) {
+        curr.clear();
+
+        std::vector<float> prob = std::vector<float>(num_classes, 0.f);
+        softmax(it, it + num_classes, prob);
+
+        for (int candidate_num = 0; candidate_num < static_cast<int>(last.size()); candidate_num++) {
+            float prob_not_blank = 0.f;
+            auto candidate = last[candidate_num];
+            std::string candidate_sentence = candidate.sentence;
+            if (candidate_sentence != "") {
+                int n = static_cast<int>(candidate_sentence.back());
+                prob_not_blank = candidate.prob_blank * prob[n];
+            }
+            float prob_blank = candidate.prob() * prob[(num_classes - 1)];
+
+            auto check_res = std::find_if(curr.begin(), curr.end(), [candidate_sentence](const BeamElement& n) {
+                return n.sentence == candidate_sentence;
+            });
+            if (check_res == std::end(curr)) {
+                curr.push_back(BeamElement{candidate.sentence, prob_blank, prob_not_blank});
+            } else {
+                auto index = std::distance(curr.begin(), check_res);
+                curr[index].prob_not_blank  += prob_not_blank;
```

Comment:
One more comment (I made it to this part of code earlier, but it is outdated)
I recommend to add here the following check:
```cpp
if curr[index].prob_blank != 0.f
    throw std::logic_error("Probability that the last char in CTC-sequence is the special blank char must be zero here");
```
This is an additional check on the consistency of the algorithm.

See detailed explanation why the check is required in my comments
* here https://github.com/opencv/open_model_zoo/pull/768#discussion_r368610795
* and here https://github.com/opencv/open_model_zoo/pull/768#discussion_r368614569